packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -1116,24 +1128,49 @@ export class LocalFluidDataStoreContextBase extends FluidDataStoreContext {
 			this.setInMemoryRoot();
 		}
 		this.createProps = props.createProps;
-		this.attachListeners();
 	}
 
-	private attachListeners(): void {
-		this.once("attaching", () => {
-			assert(
-				this.attachState === AttachState.Detached,
-				0x14d /* "Should move from detached to attaching" */,
-			);
-			this._attachState = AttachState.Attaching;
-		});
-		this.once("attached", () => {
-			assert(
-				this.attachState === AttachState.Attaching,
-				0x14e /* "Should move from attaching to attached" */,
-			);
-			this._attachState = AttachState.Attached;
-		});
+	public setAttachState(attachState: AttachState.Attaching | AttachState.Attached): void {
+		switch (attachState) {
+			case AttachState.Attaching:
+				assert(
+					this.attachState === AttachState.Detached,
+					0x14d /* "Should move from detached to attaching" */,
+				);
+				this._attachState = AttachState.Attaching;
+				if (this.channel?.setAttachState) {
+					this.channel.setAttachState(attachState);
+				} else if (this.channel) {
+					// back-compat! To be removed in the future
+					this.emit("attaching");
+				}
+				break;
+			case AttachState.Attached:
+				assert(
+					this.attachState === AttachState.Attaching,
+					0x14e /* "Should move from attaching to attached" */,
+				);
+				this._attachState = AttachState.Attached;
+				this.channel?.setAttachState?.(attachState);
+				if (this.channel?.setAttachState) {
+					this.channel.setAttachState(attachState);
+				} else if (this.channel) {
+					// back-compat! To be removed in the future
+					this.emit("attached");
+				}
+
+				// Previous code was subscribing to "attached" using once().
+				// We can get called into here twice, as result of both container and data store being attached, if
+				// those processes overlapped, for example, in a flow like that one:
+				// 1. cotnainer attach started
+				// 2. data store attachment started
+				// 3. container attached
+				// 4. data store attached.
+				this.setAttachState = () => {};
```

Comment:
This seems awkward. Instead, we can do the following I believe which is cleaner IMO:
1. If `attachState == AttachState.Attaching`, assert that `this.attachState != AttachState.Attached`. Then if the current state is attaching, just return.
2. If `attachState == AttachState.Attached`, assert that `this.attachState != AttachState.Detached`. Then if the current state is attached, just return.

Reply:
This will take care of the scenario you mentioned here. Because the flow can never go from attaced -> attaching.

Reply:
Actually for # 1 above, we can keep the same assert because a data store cannot move to attaching state twice. Only attached can happen twice.

Reply:
Changed it - take a look